File: libtest/src/test/java/org/jetbrains/kotlinx/lincheck/tests/juc/hash_map/HashMapTest.java
package org.jetbrains.kotlinx.lincheck.tests.juc.hash_map;

/*
 * #%L
 * libtest
 * %%
 * Copyright (C) 2015 - 2018 Devexperts, LLC
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Lesser Public License for more details.
 * 
 * You should have received a copy of the GNU General Lesser Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/lgpl-3.0.html>.
 * #L%
 */

import org.jetbrains.annotations.NotNull;
import org.jetbrains.kotlinx.lincheck.LinChecker;
import org.jetbrains.kotlinx.lincheck.annotations.Operation;
import org.jetbrains.kotlinx.lincheck.annotations.Param;
import org.jetbrains.kotlinx.lincheck.paramgen.IntGen;
import org.jetbrains.kotlinx.lincheck.strategy.stress.StressCTest;
import org.jetbrains.kotlinx.lincheck.verifier.VerifierState;
import org.junit.Test;

import java.util.HashMap;
import java.util.Map;

@StressCTest
@Param(name = "key", gen = IntGen.class)
@Param(name = "value", gen = IntGen.class, conf = "distinct")
public class HashMapTest extends VerifierState {
    private Map<Integer, Integer> m = new HashMap<>();

    @Operation(params = {"key", "value"})
    public Integer put(Integer key, Integer value) {
        return m.put(key, value);
    }

    @Operation
    public Integer get(@Param(name = "key") Integer key) {
        return m.get(key);
    }

    @Test(expected = AssertionError.class)
    public void test() throws Exception {
        LinChecker.check(HashMapTest.class);
    }

    @Override
    protected Object extractState() {
        return m;
    }
}

File: src/native/test/verifier/linearizability/HashMapTest.kt
/*-
 * #%L
 * Lincheck
 * %%
 * Copyright (C) 2019 - 2020 JetBrains s.r.o.
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Lesser Public License for more details.
 * 
 * You should have received a copy of the GNU General Lesser Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/lgpl-3.0.html>.
 * #L%
 */
package verifier.linearizability

import AbstractLincheckStressTest
import org.jetbrains.kotlinx.lincheck.*
import org.jetbrains.kotlinx.lincheck.paramgen.*
import org.jetbrains.kotlinx.lincheck.strategy.*

class HashMapTest : AbstractLincheckStressTest<HashMapTest>(IncorrectResultsFailure::class, UnexpectedExceptionFailure::class) {
    private val m = HashMap<Int, Int>()

    fun put(key: Int, value: Int): Int? = m.put(key, value)

    operator fun get(key: Int?): Int? = m[key]

    override fun <T : LincheckStressConfiguration<HashMapTest>> T.customize() {
        initialState { HashMapTest() }

        val keyGen = IntGen("")

        operation(IntGen(""), keyGen, HashMapTest::put)
        operation(keyGen, HashMapTest::get)
    }

    override fun extractState(): Any = m
}

